### PR TITLE
CSV Settlement Downloads

### PR DESCRIPTION
**Pull request by 00eadeyemi**
## Features

- CSV downloads for settlement reports
- Custom file names for CSV downloads.

[Delivers#118827323](https://www.pivotaltracker.com/story/show/118827323)

### CODE REVIEW

File: app/Http/Middleware/Download.php [hunk 1]
```diff
@@ -37,8 +37,9 @@ public function handle($request, Closure $next)
         $response = $next($request);
 
         if ($request->get('download') && array_key_exists('api_data', $response->original->getData())) {
-
+//           $customFilename =  array_key_exists('custom_filename', $response->original->getData())
```
Comment (wnowicki) [inline app/Http/Middleware/Download.php]: `56%` of this comment could be valid code. Did you maybe forget this after debugging?

File: resources/views/settlements/settlement_report.blade.php [hunk 2]
```diff
@@ -2,7 +2,12 @@
 
 @section('content')
 
-    <h1>SETTLEMENT REPORT</h1>
+    <h1>
+        SETTLEMENT REPORT
+        <div class="btn-group pull-right">
+            <a href="{!! Request::url() !!}/?download=csv&limit=5000" class="btn btn-default"><span class="glyphicon glyphicon-save" aria-hidden="true"></span> Download CSV</a>
```
Comment (wnowicki) [inline resources/views/settlements/settlement_report.blade.php]: `&amp;`

File: app/Http/Middleware/Download.php [hunk 3]
```diff
@@ -37,8 +37,9 @@ public function handle($request, Closure $next)
...
             switch ($request->get('download')) {
+
```
Comment (wnowicki) [inline app/Http/Middleware/Download.php]: ??

**Comment by 00eadeyemi** [comment]
@wnowicki All amendments done.

File: app/Http/Middleware/Download.php [hunk 4]
```diff
@@ -38,11 +38,14 @@ public function handle($request, Closure $next)
 
         if ($request->get('download') && array_key_exists('api_data', $response->original->getData())) {
 
+            $hasCustomFilename =  array_key_exists('export_custom_filename', $response->original->getData());
+            $filename = ($hasCustomFilename == true ? $response->original->getData()['export_custom_filename'] : 'export_' . date('Y-m-d_Hi'));
```
Comment (wnowicki) [inline app/Http/Middleware/Download.php]:
```php
if (array_key_exists('export_custom_filename', $response->original->getData()) && !is_null(array_key_exists('export_custom_filename', $response->original->getData()))) {
    $filename = $response->original->getData()['export_custom_filename'];
} else {
    $filename = 'export_' . date('Y-m-d_Hi');
}
```

- just a suggestion, reads better
- and you need to check if is not empty or null

**Comment by 00eadeyemi** [comment]
@wnowicki changes done and tested.

File: app/Http/Middleware/Download.php [hunk 5]
```diff
@@ -38,11 +38,17 @@ public function handle($request, Closure $next)
 
         if ($request->get('download') && array_key_exists('api_data', $response->original->getData())) {
 
+            if (array_key_exists('export_custom_filename', $response->original->getData()) && !is_null($filename = $response->original->getData()['export_custom_filename'])) {
```
Comment (wnowicki) [inline app/Http/Middleware/Download.php]: `!is_null($filename = $response->original->getData()['export_custom_filename'])` you don't need to assign it here and then after

**Comment by 00eadeyemi** [comment]
@wnowicki code corrected

**Pull request merged**